### ISSUE TITLE
RuntimeEnvironment.GetRuntimeInterfaceAsObject is not supported

### ISSUE THREAD

**Issue by kevingosse** [dotnet/winforms#247]
Running the Microsoft.DotNet.Analyzers.Compatibility, I found one occurence of API that is not supported on .NET Core. 
In `ComponentManagerBroker` we call [`RuntimeEnvironment.GetRuntimeInterfaceAsObject`](https://github.com/dotnet/winforms/blob/master/src/System.Windows.Forms/src/System/Windows/Forms/ComponentManagerBroker.cs#L192). This method [throws a PlatformNotSupportedException](https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/System.Runtime.InteropServices/src/System/Runtime/InteropServices/RuntimeEnvironment.cs#L28).

I wasn't actually capable of writing a repro. This code is hit [only in the case where there is a "native component manager" ](https://github.com/dotnet/winforms/blob/master/src/System.Windows.Forms/src/System/Windows/Forms/Application.cs#L2488). Since we're trying to retrieve a [`IMsoComponentManager`](https://docs.microsoft.com/en-us/previous-versions/office/developer/office-2010/gg702934(v=office.14)), I'm guessing this is specific to hosting scenarios in Microsoft Office.

**Comment by weltkante** [comment]
Besides Office this is also used in Visual Studio (called `IOleComponentManager` but its the same interface) and in VB6.

Note that the interface declaration used by WinForms is actually wrong, but apparently it slips through the cracks and doesn't crash anything. (Arguments that should be pointers are commented out and specified as `int`, which is wrong for 64 bit or when the pointer is to a COM object which needs reference counting.)

I don't know what the plans for .NET Core + Office are. If you can define a ComVisible class in .NET Core and register it, then Office would be able to call into your addin, and if you make usage of WinForms the `IMsoComponentManager` integration is vital to coordinate the message loop and form modailty with Office.

Any modifications on the implementation probably need to be coordinated with Office since they are the major consumer/provider of this interface, and its semi-undocumented. It was included in the VS 2010 SDK (which included Office headers for this interface) and in the Office documentation you linked, but Office 2010 documentation is outdated and I didn't see newer documentation.

The VS version of the interface probably can't be used authoritively because VS has no 64 bit implementation but Office does.

**Comment by merriemcgaw** [comment]
Issue moved to [dotnet/corefx #36488](https://github.com/dotnet/corefx/issues/36488) via [**ZenHub**](https://www.zenhub.com/)

**Comment by weltkante** [comment]
@merriemcgaw probably should reopen as corefx doesn't support multiple AppDomains and the unsupported API is used to determine the default AppDomain. You can assume you always are in the default AppDomain and refactor the code accordingly.

Since coreclr can now host COM objects nothing prevents users from attempting to implement an Office Addin in .NET Core and showing WinForms UI from it - they'll run in above error for now, which is probably ok to communicate its unsupported.

If you require tests during AppDomain refactoring (which I'd strongly advise since the current interface definitions appear to be wrong in some places) you'd ideally bring in someone from the Office team to help building proper tests, as said above they maintain the major relevant implementation of that interface and the only one I'm aware of which has a working 64 bit implementation. Alternatively Visual Studio also has an implementation and may be able to assist in providing tests.

**Comment by merriemcgaw** [comment]
Fair point.. We will reopen for now. I'll find an office contact to help us test this scenario.

**Comment by RussKie** [comment]
/cc: @JeremyKuhne - if you have a spare cycle or two, could you cast your eye... 
may be give a hint on how to correct `IMsoComponentManager` definitions

**Comment by JeremyKuhne** [comment]
I've actually rewritten some of them to be correct. I need to look at the VS implementations and make sure they're actually using the right definitions. I'll poke a bit into this later today.

**Comment by weltkante** [comment]
I still highly recommend checking with the Office team, they are the main consumers/providers of it besides VS, and as far as I know the only ones with a 64 bit implementation which is actually used in practice (VS still being 32 bit only)

**Comment by weltkante** [comment]
Since Office is not interested into coordinating usage of ComponentManager API (see #1763) I opened #3536 for discussion of alternatives. I prefer removal or disabling of ComponentManager integration since WPF seems to do fine without it. Not starting the discussion here because this issue can be fixed independently before discussion in #3536 yields results, but if consensus is achieved first then of course both issues can be resolved together.
